### ISSUE TITLE
[Bug] Image viewports for SpriteEntities seem to be a little off

### ISSUE THREAD

**Issue by SimonVreman** [han-yaeger/yaeger#194]
This has been bugging me since starting my project but I just can't quite put my finger on it. It seems that the viewports that are generated for SpriteEntities in `SpriteAnimationDelegate` are off by half a pixel. I don't know why it happens.

<img src=https://user-images.githubusercontent.com/35310171/120235420-4119ff00-c25a-11eb-9abe-124430fde529.png width=100>
<img src=https://user-images.githubusercontent.com/35310171/120234912-34e17200-c259-11eb-859f-9099e58e1f94.png width=100>

If you zoom in you notice that in the second image, the feet of a different frame of the sheet are showing, while this is a pixel perfect grid sheet. How I worked around this is by compensating for this when creating the viewports. This isn't happening because of resampling or scaling as I'm rendering the images at their own resolution. My dirty workaround:

```java
private void addViewPort(final int row, final int column, final double frameWidth, final double frameHeight) {
    viewports.add(new Rectangle2D(column * frameWidth + 0.5, row * frameHeight + 0.5, frameWidth, frameHeight));
}
```

In which the only difference is the two +0.5's.

**Comment by meronbrouwer** [comment]
@SimonVreman Could you provide some more info on this one? You crossed-out some sentences and I'm not sure what the problem now is.

**Comment by SimonVreman** [comment]
My bad, it was a mess. I rewrote it and added another image for clarity.

**Comment by meronbrouwer** [comment]
There is some rounding-off in the calculation of the viewports, so I guess they could cause these problems. Since other issues have a higher prio, this one will have to wait. But I'll investigate this one after those.

**Comment by SimonVreman** [comment]
I agree this is low priority. In the example given here it is not a rounding error, the calculations are correct confirmed by debugging.

**Comment by meronbrouwer** [comment]
Peculiar. When the time is right, this bug will be crushed.

**Comment by SimonVreman** [comment]
I've took some extra time to look into this issue so I will share my findings which might be useful in the future.

First of all, this issue does not happen when the spritesheet is not scaled and its location is a round pixel (e.g. x = 15 and y = 40 instead of x = 15.1432 and y = 40.315). When scaling the sheet, the interpolation is sometimes causing pixels that actually don't belong to the frame but to the next one to also be taken into calculation causing them to have a visible effect on the end result. (see the image, where the blue is darkened because red pixels are used by the bilinear interpolation) When shifting by a fraction of a pixel this is also visible for similar reasons. Something like nearest-neighbour interpolation would fix this, but it's really ugly so it just creates a bigger problem in return. (if it's even possible to implement here)

Using a viewport is a good solution when not scaling, but the interpolation issues it causes are probably just too much to deal with. I think the only way to deal with this for good is to actually split up the images into different images in memory, before scaling to the user preferred size. This would also fix the issue with shifting by fractions of a pixel, allowing the smooth animations that pixel-fractions make possible to stay.

I understand that this is very confusing so I'll try to clear it up with some examples.
![weird3](https://user-images.githubusercontent.com/35310171/127557094-8f6c645b-728d-4f32-969f-cc697a4c2ab0.png)
The top row is my sprite sheet. This appear exactly as shown when not fraction-shifting or scaling it. When I shift by a fraction of a pixel, the bottom is the result. Obviously, when shown together this looks right, the boxes appear to have shifted a bit to the right because the red is mixing a little with the blue. Which is why this is the behavior JavaFX uses. However if you now set a viewport to only show the right half of the sheet, there's a weird dark line on the left side of the light blue box.

This is solved with the separate image method as the bilinear interpolation algorithm just makes the left side of the light blue box a bit transparent because there's nothing to the left of the box. 

To conclude, I'd recommend to just leave this be for now as it's not that big of a deal.
